beacon-chain/sync/initial-sync/blocks_fetcher_utils.go

```diff
@@ -226,37 +233,51 @@ func (f *blocksFetcher) findForkWithPeer(ctx context.Context, pid peer.ID, slot
 	// Request blocks starting from the first non-empty slot.
 	req := &p2ppb.BeaconBlocksByRangeRequest{
 		StartSlot: nonSkippedSlot,
-		Count:     uint64(slotsPerEpoch.Mul(2)),
+		Count:     reqCount,
 		Step:      1,
 	}
 	blocks, err := f.requestBlocks(ctx, req, pid)
 	if err != nil {
 		return nil, fmt.Errorf("cannot fetch blocks: %w", err)
 	}
+	if len(blocks) == 0 {
+		return nil, errNoAlternateBlocks
+	}
+	first := blocks[0]
+	if !f.chain.HasBlock(ctx, first.Block().ParentRoot()) {
+		// Backtrack on a root, to find a common ancestor from which we can resume syncing.
+		fork, err := f.findAncestor(ctx, pid, first)
+		if err != nil {
+			return nil, fmt.Errorf("failed to find common ancestor: %w", err)
+		}
+		return fork, nil
+	}
 
 	// Traverse blocks, and if we've got one that doesn't have parent in DB, backtrack on it.
-	for i, block := range blocks {
+	for i := 1; i < len(blocks); i++ {
+		block := blocks[i]
 		parentRoot := block.Block().ParentRoot()
-		if !f.chain.HasBlock(ctx, parentRoot) {
-			log.WithFields(logrus.Fields{
-				"peer": pid,
-				"slot": block.Block().Slot(),
-				"root": fmt.Sprintf("%#x", parentRoot),
-			}).Debug("Block with unknown parent root has been found")
-			// Backtrack only if the first block is diverging,
-			// otherwise we already know the common ancestor slot.
-			if i == 0 {
-				// Backtrack on a root, to find a common ancestor from which we can resume syncing.
-				fork, err := f.findAncestor(ctx, pid, block)
-				if err != nil {
-					return nil, fmt.Errorf("failed to find common ancestor: %w", err)
-				}
-				return fork, nil
-			}
-			return &forkData{peer: pid, blocks: blocks}, nil
+		if f.chain.HasBlock(ctx, parentRoot) {
+			continue
+		}
+		log.WithFields(logrus.Fields{
+			"peer": pid,
+			"slot": block.Block().Slot(),
+			"root": fmt.Sprintf("%#x", parentRoot),
+		}).Debug("Block with unknown parent root has been found")
+		altBlocks, err := validROBlocks(blocks[i-1:])
```

Comment:
Don't understand this, why will we want to fetch blobs for these blocks ? Don't they have unknown parents

Reply:
My understanding of this code path is that when we detect get stuck syncing blocks, we look for peers with a different head and then try to backtrack to a common ancestor so that we can attempt to sync the alternate branch. In this change I am just trying to clean up the existing implementation a little because the loop was written in a super weird way and add blob support. Maybe we could walk through this code more carefully to review together in a video call.

Reply:
I'm going to add comments to this method to clarify why we're doing this.